exec/vector/src/main/java/org/apache/drill/exec/record/metadata/DictBuilder.java

```diff
@@ -0,0 +1,198 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.drill.exec.record.metadata;
+
+import org.apache.drill.common.types.TypeProtos;
+import org.apache.drill.exec.record.MaterializedField;
+import org.apache.drill.exec.vector.complex.DictVector;
+
+public class DictBuilder implements SchemaContainer {
+
+  private final SchemaContainer parent;
+  private final TupleBuilder tupleBuilder = new TupleBuilder();
+  private final String memberName;
+  private final TypeProtos.DataMode mode;
+
+  public DictBuilder(String memberName, TypeProtos.DataMode mode) {
+    this(null, memberName, mode);
+  }
+
+  public DictBuilder(SchemaContainer parent, String memberName, TypeProtos.DataMode mode) {
+    this.parent = parent;
+    this.memberName = memberName;
+    this.mode = mode;
+  }
+
+  @Override
+  public void addColumn(ColumnMetadata column) {
+    assert DictVector.fieldNames.contains(column.name());
+    tupleBuilder.addColumn(column);
+  }
+
+  public DictBuilder addKey(TypeProtos.MajorType type) {
+    return add(MaterializedField.create(DictVector.FIELD_KEY_NAME, type));
+  }
+
+  public DictBuilder addValue(TypeProtos.MajorType type) {
+    return add(MaterializedField.create(DictVector.FIELD_VALUE_NAME, type));
+  }
+
+  public DictBuilder add(MaterializedField col) {
+    assert DictVector.fieldNames.contains(col.getName());
+    tupleBuilder.add(col);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, TypeProtos.DataMode mode) {
+    tupleBuilder.add(DictVector.FIELD_KEY_NAME, type, mode);
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, TypeProtos.DataMode mode) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type, mode);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type) {
+    tupleBuilder.add(DictVector.FIELD_KEY_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, int width) {
+    tupleBuilder.add(DictVector.FIELD_KEY_NAME, type, width);
+    return this;
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int width) {
+    tupleBuilder.add(DictVector.FIELD_VALUE_NAME, type, width);
+    return this;
+  }
+
+  public DictBuilder addKey(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_KEY_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addValue(TypeProtos.MinorType type, int precision, int scale) {
+    return addDecimal(DictVector.FIELD_VALUE_NAME, type, TypeProtos.DataMode.REQUIRED, precision, scale);
+  }
+
+  public DictBuilder addNullable(String name, TypeProtos.MinorType type) {
```

Comment:
Yeah, all these forms make sense when only working with the top-level row, but become tedious when repeated or all these specialized builders...